Source/WTF/wtf/VariantListOperations.h

```diff
@@ -0,0 +1,527 @@
+}
+
+template<typename V> auto VariantListOperations<V>::readTag(std::span<std::byte> buffer) -> VariantListTag
+{
+    VariantListTag value;
+    memcpy(&value, buffer.data(), sizeof(VariantListTag));
+    return value;
+}
+
+template<typename V> auto VariantListOperations<V>::readTag(std::span<const std::byte> buffer) -> VariantListTag
+{
+    VariantListTag value;
+    memcpy(&value, buffer.data(), sizeof(VariantListTag));
+    return value;
+}
+
+template<typename V> auto VariantListOperations<V>::writeTag(VariantListTag tag, std::span<std::byte> buffer) -> std::span<std::byte>
+{
+    memcpy(buffer.data(), &tag, sizeof(VariantListTag));
+    return buffer.subspan(sizeof(VariantListTag));
+}
+
+template<typename V> template<typename T> T& VariantListOperations<V>::readValue(std::span<std::byte> buffer)
+{
+    constexpr auto tagFromType = alternativeIndexV<T, V>;
+    ASSERT_UNUSED(tagFromType, tagFromType == readTag(buffer));
+
+    return reinterpretCastSpanStartTo<T>(alignBufferForValue<T>(buffer.subspan(sizeof(VariantListTag))));
+}
+
+template<typename V> template<typename T> const T& VariantListOperations<V>::readValue(std::span<const std::byte> buffer)
+{
+    constexpr auto tagFromType = alternativeIndexV<T, V>;
+    ASSERT_UNUSED(tagFromType, tagFromType == readTag(buffer));
+
+    return reinterpretCastSpanStartTo<T>(alignBufferForValue<T>(buffer.subspan(sizeof(VariantListTag))));
+}
+
+template<typename V> template<typename Arg> auto VariantListOperations<V>::writeValue(Arg&& value, std::span<std::byte> buffer) -> std::span<std::byte>
+{
+    using T = std::decay_t<Arg>;
+
+    new (NotNull, buffer.data()) T(std::forward<Arg>(value));
+    return buffer.subspan(sizeof(T));
+}
+
+template<typename V> template<size_t I, typename F> constexpr ALWAYS_INLINE decltype(auto) VariantListOperations<V>::visitTypeForTag(VariantListTag tag, F&& f)
+{
+    constexpr auto size = std::variant_size_v<V>;
+
+#define WTF_VARIANT_LIST_VISIT_CASE_COUNT 32
+#define WTF_VARIANT_LIST_VISIT_CASE(N, D) \
+        case I + N:                                                                                 \
+        {                                                                                           \
+            if constexpr (I + N < size) {                                                           \
+                return f.template operator()<std::variant_alternative_t<I + N, V>>();               \
+            } else {                                                                                \
+                WTF_UNREACHABLE();                                                                  \
+            }                                                                                       \
+        }                                                                                           \
+
+    switch (tag) {
+        WTF_VARIANT_LIST_VISIT_CASE(0, WTF_VARIANT_LIST_VISIT_CASE_COUNT)
```

Comment:
An idiom like this seems like it needs the comment that explains what its intentions are.